### PR TITLE
Threat intel monitor implementation

### PR DESCRIPTION
**Pull request by eirsep**
### Description
Threat intel monitor is a job that runs periodically to scan the configured customer data for malicious indicators. It's corpus of malicious indicators is a list of system indices containing threat intel data curated from the configured threat intelligence source configs.

We refer threat intel source configs that are in **AVAILABLE or REFRESHING** state.
we fetch the ioc system indices mentioned in these TIF source configs for corpus of indicators.
Threat intel monitor does the following steps when it runs:
- fetch configs
- get ioc indices for each ioc type
- fetch documents since last seen seq_no for the assigned data sources
- extract iocs from docs from configured fields for each ioc type
- search for extracted ioc list with terms query in batches of MAX_TERMS_SETTING against the ioc indices
- create Findings fi there are malicious iocs in data matched against threat intel sysytem indices

TODO: pending alert triggers and notifications
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestGetIocFindingsAction.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.List;
+
+import static java.util.Collections.singletonList;
+import static org.opensearch.rest.RestRequest.Method.GET;
+
+public class RestGetIocFindingsAction extends BaseRestHandler {
+
+    @Override
+    public String getName() {
+        return "get_ioc_findings_action_sa";
```
Comment (AWSHurneyt) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestGetIocFindingsAction.java:30]: Is this PR intended to have some assets from this PR?
https://github.com/opensearch-project/security-analytics/pull/1093/files#diff-c79121f958035e722c3a1b9f151ea28d5d9905080d907b8f53ae12005c0d100fR30

File: src/main/java/org/opensearch/securityanalytics/threatIntel/transport/TransportGetIocFindingsAction.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import org.opensearch.securityanalytics.threatIntel.iocscan.dao.IocFindingService;
+import org.opensearch.securityanalytics.transport.SecureTransportAction;
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+import java.time.Instant;
+import java.util.List;
+
+public class TransportGetIocFindingsAction extends HandledTransportAction<GetIocFindingsRequest, GetIocFindingsResponse> implements SecureTransportAction {
```
Comment (AWSHurneyt) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/transport/TransportGetIocFindingsAction.java:43]: Same comment as above; looks like this PR has some assets from PR https://github.com/opensearch-project/security-analytics/pull/1093

File: src/test/java/org/opensearch/securityanalytics/resthandler/ThreatIntelMonitorRestApiIT.java [hunk 3]
```diff
@@ -27,7 +39,67 @@
...
+            indexTifSourceConfig(num, configId, iocIndexName, i);
+            for (int i1 = 0; i1 < iocVals.size(); i1++) {
+                indexIocs(iocVals, iocIndexName, i1, configId);
+            }
+        }
+    }
+
+    private void indexIocs(List<String> iocVals, String iocIndexName, int i1, String configId) throws IOException {
+        String iocId = iocIndexName + i1;
+        STIX2IOC stix2IOC = new STIX2IOC(
```
Comment (AWSHurneyt) [inline src/test/java/org/opensearch/securityanalytics/resthandler/ThreatIntelMonitorRestApiIT.java:55]: For reference, I have some static functions for creating random IOCs in this helper class.
https://github.com/opensearch-project/security-analytics/blob/feature/threat_intel/src/test/java/org/opensearch/securityanalytics/util/STIX2IOCGenerator.java#L77

File: src/main/java/org/opensearch/securityanalytics/threatIntel/transport/TransportGetIocFindingsAction.java (hunk 2 above)
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/transport/TransportGetIocFindingsAction.java:51]: i am blocked on the PR so i integrated.. will rebased that PR's commits.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestGetIocFindingsAction.java (hunk 1 above)
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/resthandler/RestGetIocFindingsAction.java:30]: i am blocked on the PR so i integrated.. will rebased that PR's commits.

File: src/main/java/org/opensearch/securityanalytics/threatIntel/model/SATIFSourceConfig.java [hunk 4]
```diff
@@ -268,6 +268,9 @@ public static SATIFSourceConfig parse(XContentParser xcp, String id, Long versio
                 case NAME_FIELD:
                     name = xcp.text();
                     break;
+                case VERSION_FIELD:
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/model/SATIFSourceConfig.java:271]: i am blocked on the actual fix so i made an adhoc.. will rebased from feature branch

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/SaIoCScanService.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+
+public class SaIoCScanService extends IoCScanService<SearchHit> {
+
+    private static final Logger log = LogManager.getLogger(SaIoCScanService.class);
+    public static final int MAX_TERMS = 65536; //make ioc index setting based. use same setting value to create index
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/SaIoCScanService.java:42]: ack

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    abstract void matchAgainstThreatIntelAndReturnMaliciousIocs(
+            Map<String, Set<String>> iocsPerType,
+            Monitor monitor,
+            BiConsumer<List<STIX2IOC>, Exception> callback,
+            Map<String, List<String>> iocTypeToIndices);
+
+    /**
+     * For each doc, we extract the list of
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java:85]: done

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        try {
+            List<Data> data = iocScanContext.getData();
+            if (data.isEmpty()) {
+                scanCallback.accept(Collections.emptyList(), null);
+                return;
+            }
+            Monitor monitor = iocScanContext.getMonitor();
+
+            long startTime = System.currentTimeMillis();
+            // log.debug("beginning to scan IoC's")
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java:41]: removed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dto/IocScanContext.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package org.opensearch.securityanalytics.threatIntel.iocscan.dto;
+
+import org.opensearch.commons.alerting.model.Monitor;
+import org.opensearch.commons.alerting.model.MonitorMetadata;
+import org.opensearch.securityanalytics.threatIntel.model.monitor.ThreatIntelInput;
+
+import java.util.List;
+import java.util.Map;
+
+public class IocScanContext<Data> {
```
Comment (jowg-amazon) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dto/IocScanContext.java:10]: Does IocScanContext need serialization/deserialization?

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+                }
+            }
+            saveIocs(iocFindings, callback, monitor);
+        } catch (Exception e) {
+            log.error(() -> new ParameterizedMessage("Failed to create ioc findinges due to unexpected error {}", iocScanContext.getMonitor().getId()), e);
+            callback.accept(null, e);
+        }
+    }
+
+    abstract void saveIocs(List<IocFinding> iocs, BiConsumer<List<IocFinding>, Exception> callback, Monitor monitor);
```
Comment (jowg-amazon) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java:176]: `saveIocFindings` might be a better name to differentiate saving the actual iocs

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dto/IocScanContext.java (hunk 8 above)
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dto/IocScanContext.java:11]: no
just dto between classes - single node operation too

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java (hunk 9 above)
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/service/IoCScanService.java]: fixed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                    r -> {
+                        log.debug("{} index created", getIndexName());
+                        listener.onResponse(null);
+                    }, e -> {
+                        if (e instanceof ResourceAlreadyExistsException) {
+                            log.debug("index {} already exist", getIndexMapping());
+                            listener.onResponse(null);
+                            return;
+                        }
+                        log.error("Failed to create security analytics threat intel job index", e);
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java:219]: fixed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+    public void searchEntities(SearchSourceBuilder searchSourceBuilder, final ActionListener<SearchResponse> actionListener) {
+        SearchRequest searchRequest = new SearchRequest()
+                .source(searchSourceBuilder)
+                .indices(getIndexName());
+        client.search(searchRequest, actionListener);
+    }
+
+    public void createIndexIfNotExists(final ActionListener<Void> listener) {
+        // check if job index exists
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java:201]: fixed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/action/GetIocFindingsAction.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.securityanalytics.threatIntel.action;
+
+import org.opensearch.action.ActionType;
+
+public class GetIocFindingsAction extends ActionType<GetIocFindingsResponse> {
+
+    public static final GetIocFindingsAction INSTANCE = new GetIocFindingsAction();
+    public static final String NAME = "cluster:admin/opensearch/securityanalytics/ioc/findings/get";
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/action/GetIocFindingsAction.java:12]: for rest api naming plural

File: src/main/java/org/opensearch/securityanalytics/logtype/BuiltinLogTypeLoader.java [hunk 13]
```diff
@@ -69,8 +70,11 @@ public void ensureLogTypesLoaded() {
     private List<LogType> loadBuiltinLogTypes() throws URISyntaxException, IOException {
         List<LogType> logTypes = new ArrayList<>();
 
-        final String url = Objects.requireNonNull(BuiltinLogTypeLoader.class.getClassLoader().getResource(BASE_PATH)).toURI().toString();
+        String pathurl = Paths.get(BuiltinLogTypeLoader.class.getClassLoader().getResource(BASE_PATH).toURI()).toString();
 
+        final String url = Objects.requireNonNull(BuiltinLogTypeLoader.class.getClassLoader().getResource(BASE_PATH)).toURI().toString();
+        logger.error("SASHANK Path url is {}", pathurl);
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/logtype/BuiltinLogTypeLoader.java:76]: removed thanks

File: src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java [hunk 14]
```diff
@@ -330,72 +334,6 @@ public void deleteTIFSourceConfig(
         ));
     }
 
-    public void deleteAllOldIocIndices(List<String> indicesToDelete) {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigService.java:333]: reverted

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/IocFindingService.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+            }
+        } catch (IOException e) {
+            log.error("Failed to get the threat intel ioc match index mapping", e);
+            throw new SecurityAnalyticsException("Failed to get the threat intel ioc match index mapping", RestStatus.INTERNAL_SERVER_ERROR, e);
+        }
+    }
+
+    /**
+     * Index name: .opensearch-sap-iocmatch
+     * Mapping: /mappings/ioc_match_mapping.json
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/IocFindingService.java]: updated

File: src/main/java/org/opensearch/securityanalytics/action/GetCorrelationAlertsAction.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.securityanalytics.action;
+
+import org.opensearch.action.ActionType;
+
+public class GetCorrelationAlertsAction extends ActionType<GetCorrelationAlertsResponse> {
+
+    public static final GetCorrelationAlertsAction INSTANCE = new GetCorrelationAlertsAction();
+    public static final String NAME = "cluster:admin/opensearch/securityanalytics/correlationAlerts/get";
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/action/GetCorrelationAlertsAction.java:12]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+                    listener.onFailure(e);
+                }
+        ));
+    }
+
+    public void indexCorrelationAlert(CorrelationAlert correlationAlert, TimeValue indexTimeout, ActionListener<IndexResponse> listener) {
+        // Convert CorrelationAlert to a map
+        try {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+            builder.field("correlated_finding_ids", correlationAlert.getCorrelatedFindingIds());
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java:125]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+                .source(searchSourceBuilder);
+
+        // Execute the search request
+        client.search(searchRequest, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse searchResponse) {
+                // Iterate through the search hits
+                for (SearchHit hit : searchResponse.getHits().getHits()) {
+                    // Construct a script to update the document with the new state and acknowledgedTime
+                    // Construct a script to update the document with the new state and acknowledgedTime
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java:213]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+        // Execute the search request
+        client.search(searchRequest, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse searchResponse) {
+                // Iterate through the search hits
+                for (SearchHit hit : searchResponse.getHits().getHits()) {
+                    // Construct a script to update the document with the new state and error_message
+                    Script script = new Script(ScriptType.INLINE, "painless",
+                            "ctx._source.state = params.state; ctx._source.error_message = params.error_message",
+                            Map.of("state", Alert.State.ERROR, "error_message", "The rule associated to this Alert is deleted"));
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/correlation/alert/CorrelationAlertService.java:286]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/correlation/alert/notifications/NotificationService.java [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+        idSet.add(id);
+        GetNotificationConfigRequest getNotificationConfigRequest = new GetNotificationConfigRequest(idSet, 0, 10, null, null, new HashMap<>());
+        GetNotificationConfigResponse configResp = null;
+        NotificationsPluginInterface.INSTANCE.getNotificationConfig(client, getNotificationConfigRequest, new ActionListener<GetNotificationConfigResponse>() {
+            @Override
+            public void onResponse(GetNotificationConfigResponse getNotificationConfigResponse) {
+                if (getNotificationConfigResponse.getStatus() == RestStatus.OK) {
+                    getNotificationConfigResponse = configResp;
+                } else {
+                    logger.error("Successfully sent a notification, Notification Event: " + getNotificationConfigResponse);
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/correlation/alert/notifications/NotificationService.java:113]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/model/CorrelationRuleTrigger.java [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CorrelationRuleTrigger implements Writeable, ToXContentObject {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/model/CorrelationRuleTrigger.java:32]: not from my PR
merged main branch

File: src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigManagementService.java [hunk 22]
```diff
@@ -479,25 +479,25 @@ private void checkAndDeleteOldIocIndicesByAge(
...
-                                }
-                            }
-                            stepListener.onResponse(indicesToDelete);
-                        }, e -> {
-                            log.error("Failed to get the cluster metadata");
-                            stepListener.onFailure(e);
-                        }
-                ), indices.toArray(new String[0])
-        );
+//        log.debug("Delete old IOC indices by age");
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/service/SATIFSourceConfigManagementService.java:482]: reverted

File: src/main/java/org/opensearch/securityanalytics/model/threatintel/ThreatIntelAlert.java [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+            String errorMessage,
+            String severity,
+            String iocValue,
+            String iocType,
+            List<ActionExecutionResult> actionExecutionResults,
+            List<String> findingIds
+    ) {
+
+        this.id = id != null ? id : NO_ID;
+        this.version = version != 0 ? version : NO_VERSION;
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/model/threatintel/ThreatIntelAlert.java:90]: its primitive `long`

File: src/main/java/org/opensearch/securityanalytics/model/threatintel/IocWithFeeds.java [hunk 24]
```diff
@@ -108,4 +108,34 @@ public static IocWithFeeds parse(XContentParser xcp) throws IOException {
     public static IocWithFeeds readFrom(StreamInput sin) throws IOException {
         return new IocWithFeeds(sin);
     }
+
+    @Override
+    public int hashCode() {
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/model/threatintel/IocWithFeeds.java:113]: changed

File: src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java [hunk 25]
```diff
@@ -0,0 +1,256 @@
...
+                    }));
+
+
+        } catch (Exception e) {
+            log.error("Exception saving the threat intel source config in index", e);
+            actionListener.onFailure(e);
+        }
+    }
+
+    public void bulkIndexEntities(List<Entity> entityList,
```
[the remark this answers is not in the document]
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/threatIntel/iocscan/dao/BaseEntityCrudService.java:136]: one is for only new entities and 
one for create new and update old entities

File: src/main/java/org/opensearch/securityanalytics/threatIntel/transport/monitor/TransportGetThreatIntelAlertsAction.java [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+                        listener.onResponse(new GetThreatIntelAlertsResponse(Collections.emptyList(), 0));
+                        return;
+                    }
+                    log.error("Failed to get threat intel monitor alerts", e);
+                    listener.onFailure(e);
+                }
+        ));
+    }
+
+    private void getAlerts(List<String> monitorIds,
```
Comment (jowg-amazon) [inline src/main/java/org/opensearch/securityanalytics/threatIntel/transport/monitor/TransportGetThreatIntelAlertsAction.java:135]: not blocking but should logic be moved to a service class outside of transport?

**Pull request merged**